Pyspark/regular_patients/regular_patients.py

```diff
@@ -0,0 +1,136 @@
+    def __init__(self, input_data) -> None :
+        """Initializes the class and creates the spark session"""
+        self.spark : SparkSession = (SparkSession.builder.appName('PatientsData')
+            .config('spark.driver.extraJavaOptions',
+                    '-Dlog4j.configuration=file:regular_patients/config/log4j.properties')
+            .getOrCreate())
+        self.df_patients : DataFrame  = self.load_data(input_data)
+        super().__init__(self.spark)
+
+    def load_data(self, input_data: str) -> DataFrameReader :
+        """Loads the CSV data file, inferring the schema"""
+        return (self.spark.read
+        .format ('csv')
+        .option('header', 'true')
+        .option('inferschema', 'true')
+        .load(input_data))
+
+    def format_date(self) -> DataFrame :
+        """Takes the effective_date field and tranforms it to a date type"""
+        self.df_patients = (self.df_patients
+            .withColumn('effective_from_date_formatted',
+                        f.col('effective_from_date').cast('string'))
+            .withColumn('effective_from_date_formatted',
+                         f.lpad(f.col('effective_from_date_formatted'),  8, '0'))
+            .withColumn('effective_from_date_formatted',
+                        f.to_date(f.col('effective_from_date_formatted'), 'MMddyyyy')))
+        return self.df_patients
+
+    def filter_by_date_range(self, start_date: str, end_date: str) -> DataFrame :
+        """Filtering the data by date range"""
+        self.df_patients = self.df_patients.filter(
+        (f.col('effective_from_date_formatted') >= f.to_date(f.lit(start_date), 'yyyy-MM-dd')) &
+        (f.col('effective_from_date_formatted') <= f.to_date(f.lit(end_date), 'yyyy-MM-dd'))
+        )
+        return self.df_patients
+
+    def checking_consecutives_records(self) -> DataFrame :
+        """Identifies consecutive records for each patient based on monthly intervals. 
+        This function flags consecutive records by checking if the difference in months 
+        between consecutive entries is exactly one. It also calculates the longest sequence 
+        of consecutive records per patient"""
+        # Defining the window specification
+        win_spec = Window.partitionBy('patient_id').orderBy('effective_from_date_formatted')
+
+        # Add previous date column
+        self.df_patients = self.df_patients.withColumn(
+            'prev_date', f.lag(f.col('effective_from_date_formatted')).over(win_spec)
+        )
+        # Calculate months between current and previous record
+        self.df_patients = self.df_patients.withColumn(
+            'months_between', f.months_between(
+                f.col('effective_from_date_formatted'), f.col('prev_date')
+                )
+        )
+        # Flagging the consecutive records
+        win_spec_2 = Window.partitionBy('patient_id').orderBy('effective_from_date_formatted')
+
```

Comment:
this can be simplified 
![image](https://github.com/user-attachments/assets/33f3cfcd-bf36-4f44-8842-8b1b9aad9580)